The compiler -> a compiler (#1226)
dotnet/csharpstandard: standard/classes.md

A *class_member_declaration* is permitted to declare a member with the same name or signature as an inherited member. When this occurs, the derived class member is said to *hide* the base class member. See [§7.7.2.3](basic-concepts.md#7723-hiding-through-inheritance) for a precise specification of when a member hides an inherited member.

> An inherited member `M` is considered to be ***available*** if `M` is accessible and there is no other inherited accessible member N that already hides `M`. Implicitly hiding an inherited member is not considered an error, but it does cause a compiler to issue a warning unless the declaration of the derived class member includes a `new` modifier to explicitly indicate that the derived member is intended to hide the base member. If one or more parts of a partial declaration ([§15.2.7](classes.md#1527-partial-declarations)) of a nested type include the `new` modifier, no warning is issued if the nested type hides an available inherited member.

Reviewer: This feels like "the" would be more appropriate, personally - but I'm quite happy to go with "it's easier to be consistent".

Author: Consider this rewording:

```suggestion
An inherited member `M` is considered to be ***available*** if `M` is accessible and there is no other inherited accessible member N that already hides `M`. Implicitly hiding an inherited member is not considered an error, but a compiler shall issue a warning unless the declaration of the derived class member includes a `new` modifier to explicitly indicate that the derived member is intended to hide the base member. If one or more parts of a partial declaration ([§15.2.7](classes.md#1527-partial-declarations)) of a nested type include the `new` modifier, no warning is issued if the nested type hides an available inherited member.
```

Author: Or is "can issue a warning" better wording?

>
> *end example*

> Constants are permitted to depend on other constants within the same program as long as the dependencies are not of a circular nature. A compiler is allowed to evaluate the constant declarations in an appropriate order.

Reviewer 2: (I can't see how to re-surface @jskeet's comment in “Files Changed”, I've “unresolved” the conversation so its visible in “Conversation”.)

@jskeet asked whether this should be a language requirement.

I think a lot of the original source material *described* what “the compiler” (it was the one and only) did and was never converted into language requirements.

“A compiler is allowed to evaluate the constant declarations in an appropriate order.” implies that it is not required to do so, which is clearly not true. I think any appropriate order was probably meant?

```suggestion
Constants are permitted to depend on other constants within the same program as long as the dependencies are not of a circular nature. A compiler must evaluate the constant declarations in an appropriate order.
```

But are we saying too much? Do compiler writers require this much direction to order the dependencies? Here I might offer:

```suggestion
Constants are permitted to depend on other constants within the same program as long as the dependencies are not of a circular nature.
```

If it is felt that compiler writers need more help to figure this out we could go with something along the lines of:

```suggestion
Constants are permitted to depend on other constants within the same program as long as the dependencies are not of a circular nature. A compiler must evaluate constant declarations in an order which respects the dependencies.
```

or maybe:

```suggestion
Constants are permitted to depend on other constants within the same program as long as the dependencies are not of a circular nature. Constant declarations shall be evaluated in any order which respects any dependencies.
```

I personally lean towards the single sentence option, compiler writers can figure it out.

> Constants are permitted to depend on other constants within the same program as long as the dependencies are not of a circular nature. A compiler automatically arranges to evaluate the constant declarations in the appropriate order.

Reviewer: This second sentence doesn't sound quite like spec-ese to start with. Presumably it should be more of a *requirement* instead of a statement of what happens.

Author: reworked the sentence to permit a compiler to reorder the evaluation of constant declarations

> }
> ```
>

> > a compiler first evaluates `A.Y`, then evaluates `B.Z`, and finally evaluates `A.X`, producing the values `10`, `11`, and `12`.

Reviewer: Again, perhaps this should be "must first evaluate"? (It sounds odd with "a compiler first evaluates" to me.)

>
> *end example*

> When compiling a field-like event, a compiler automatically creates storage to hold the delegate, and creates accessors for the event that add or remove event handlers to the delegate field. The addition and removal operations are thread safe, and may (but are not required to) be done while holding the lock ([§13.13](statements.md#1313-the-lock-statement)) on the containing object for an instance event, or the `System.Type` object ([§12.8.18](expressions.md#12818-the-typeof-operator)) for a static event.

Reviewer: Possibly "a compiler shall"?
